README.md
# pyIMSafe-OSX
Basic encrypted chatroom software (for the purpose of demonstration &amp; experimentation)

This software is an example of a simple encryption algorithm. This software was written for the purpose of demonstration and experimentation and is probably not suitable to send or store sensitive information. Use at your own risk.


To run, simply open IMGui.py or IMSafe.py with python.


Python_Client/changes.txt
IMSafe 2.1 Changes
- Output is now base64 encoded

IMSafe 2.0  Changes
- Different encryption algorithm used
- Fixed decryption error bug

IMSafe 1.0 Changes
- Threaded update no longer causes crash
- Fixed dropped message bug

IMSafe 0.6 Beta Changes

- Timestamp in chat is in a smaller font
- Removed the "connect" button from the settings dialog
- Added sound options
- Added a different sound for sending than recieving

IMSafe 0.5 Beta Changes

- Better network error handling
- "http://" added to given url if necessary, instead of raising an error
- Different text highlights for screenames of other people's messages

IMSafe 0.4 Beta Changes

- Can now connect to a chatroom
- Screenames are  now highlighted in chat
- Fixed a bug where Ctrl+a would select all text and not allow you to deselect.

Server/IMSafe.php
<?php
//IMSafe Encrypted Chatroom Server Software
/*
*  Copyright (C) 2014  Patrick T. Cossette
*
*  This program is free software; you can redistribute it and/or
*  modify it under the terms of the GNU General Public License
*  as published by the Free Software Foundation; either version 2
*  of the License, or (at your option) any later version.
*  
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*  
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

//Created By: Patrick T. Cossette
//http://www.DigitalDiscrepancy.com

include 'chatdb.php';
$chat = new chatdb();

//Configure your whitelist and  blacklist  parameters here
//1 = use list
//0 = ignore list
$useWhitelist = 0;
$useBlacklist = 0;
$whitelist= array(); // Enter whitelist usernames here
$blacklist= array(); // Enter blacklist usernames here

if(!isset($_POST["username"]) || !isset($_POST["message"])){
		//This page was most likely accessed through a browser..
		?>
  		<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
        <html>
            <head>
            <title>IMSafe Server</title>
                <BODY><center><h2><br/><br/>
                Welcome to the IMSafe Server page!<br/> If you are looking to download IMSafe, you can find the latest versions <a href="http://www.fullofepicwin.com/imsafe">here</a>.<br/>
                <br/<br/></h2><h3>
                Otherwise, please refer to IMSafe's home page for help on how to use this software!</h3>
                </center>
                <br/>
                
                </BODY>
        </HTML>
     	<?php   
	}
    else{
		if(isset($_GET["room"])){
			$chatroom = $_GET["room"];
		}
		else{
			$chatroom = "Main";
		}
		
		if($_POST["message"] == "connect_request"){ 
			//A new user is attempting to connect, make sure  the username doesn't conflict with the whitelist/blacklist
			if((!$useWhitelist || in_array($_POST["username"], $whitelist)) && (!$useBlacklist || !in_array($_POST["username"], $blacklist))){
				if(!in_array($_POST["username"], $chat->getOnlineUsers($chatroom))){
					//Check  to see if the username is already in the users table, if not, add it.
					$chat->addUser($_POST["username"], $chatroom);
				}
				echo "Connection Successful";
			}
			else{
				echo "Connection Failed, ".$_POST["username"];
			}
			
		}
		else if (isset($_POST["prg"])){
			//prg represents what function the user is requesting to execute
			if ($_POST["prg"] == "list"){
				echo implode(",", $chat->getOnlineUsers($chatroom)); //Return a list of users as comma separated values
			}
			else if ($_POST["prg"] == "update"){
				//retrieve any new messages and delete messages that have  already been recieved
				$chatroomUpdate = $chat->getMessagesForUser($_POST["username"], $_POST["recipts"], $chatroom);
				echo $chatroomUpdate;
			}
			else{ 
				// Disconnect user
				$chat->disconnectUser($_POST["username"]);
			}
		}
		else{
			//The user is sending message to the chatroom
			//Currenlty 'options' only supports first person messages, but will also make private messaging  possible in the future
			$options = array();
			$options['type'] = 'normal';
			
			if(isset($_POST["options"])){
				//In the future, process the "options" string into a dictionary style array before sending it to the addMessage function
				$options['type'] = 'first_person';
			}
			
			//Send the message to the database
			$chat->addMessage($_POST["username"], $_POST["message"], $chatroom, $options);
		}
}

?>

Server/chatdb.php
<?php


//IMSafe Encrypted Chatroom Server Software
//http://www.FullOfEpicWin.com/imsafe/

/*
*  Copyright (C) 2014  Patrick T. Cossette
*
*  This program is free software; you can redistribute it and/or
*  modify it under the terms of the GNU General Public License
*  as published by the Free Software Foundation; either version 2
*  of the License, or (at your option) any later version.
*  
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*  
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

ini_set("post_max_size", "20M");

class chatdb{
	//Put your own database info here!
	public $dbname = "dbname";
	public $dbuser = "dbuser";
	public $dbpassword = "dbpassword";
	public $dbhost = "localhost";

	function __construct(){
		$connection = mysql_connect($this->dbhost, $this->dbuser, $this->dbpassword);// or die('unable to connect!');
		mysql_select_db($this->dbname) or die("Unable to access database! " . mysql_error());
		$this->create_tables();  //Configure the database for use
		$this->deleteOldUsers(); //Delete any users who have timed out due to inactivity
	}
	
	function deleteOldUsers(){
		$expiration = time()-60; //Users are automatically signed off after 5 minutes of inactivity
		$sql ="DELETE FROM users WHERE lastAction <".$expiration;
		$result = mysql_query($sql);
	}
	
	function create_tables(){
		$sql="CREATE TABLE IF NOT EXISTS users(
			username VARCHAR(255) NOT NULL,
			room VARCHAR(255) NOT NULL,
			lastAction INT
			)";
			
		$result = mysql_query($sql);

		$sql="CREATE TABLE IF NOT EXISTS messages(
			room VARCHAR(255) NOT NULL,
			id INT(11) NOT NULL auto_increment,
			fromUser VARCHAR(255) NOT NULL,
			toUser VARCHAR(255) NOT NULL,
			message BLOB NOT NULL,
			unixTime INT,
			time TIMESTAMP,
			type VARCHAR(32),
			PRIMARY KEY (id)
			 )";
			
		$result = mysql_query($sql);
	}
	
	function addUser($username, $chatroom){
		//Add a use to the databas when they first log on
		//$sql = "SELECT COUNT(*) FROM customers WHERE user_id='test'"
		
		$sql="INSERT into users (username, lastAction, room) VALUES (\"$username\", ".time().", \"$chatroom\")";
		$result = mysql_query($sql);
		if(!$result){
			die('Error: ' . mysql_error());
		}
	}
	
	function disconnectUser($username, $chatroom){
		$sql = "DELETE FROM users WHERE username=\"$username\" AND room = \"$chatroom\"";
		$result = mysql_query($sql);
	}

	function getOnlineUsers($chatroom){
		//Return an array containing all the usernames currently using the chatroom
		$users = array();
		
		$sql = "SELECT * FROM users WHERE room = \"$chatroom\"";
		$result = mysql_query($sql);
		if(!$result){
			echo "<br/>MySql Error: ".mysql_error();
		}
		
		while($row = mysql_fetch_row($result)){
			array_push($users, $row[0]);
		}
		
		return $users;
	}
	
	function addMessage($fromUser, $message, $chatroom, $options){
		//$options is a dictionary style array that will specify private messaging in the future, currenlty only used to specify first person messages
		//The database structure currently supports private messaging, but is only being used like a chatroom for now
		$users = $this->getOnlineUsers($chatroom);
	
		foreach($users as $user){
			$sql = "INSERT INTO messages (room, fromUser, toUser, message, unixTime, type) VALUES (\"$chatroom\", \"$fromUser\", \"$user\", \"$message\", ".time().",\"". $options['type'] ."\")";
			mysql_query($sql);
		}
	}
	
	function getMessagesForUser($username, $recipts, $chatroom){
	 /* Returns a list of messages for "username", and deletes messages that have a confirmed recipt.
	    Messages are NOT DELETED unless a recipt is recieved, this is to avoid dropped messages. */
		
		//First, lets update the user's lastActivity value
		$sql = "UPDATE users SET lastAction = ".time()." WHERE username=\"$username\" AND room = \"$chatroom\"";
		$result = mysql_query($sql);
	   
	    $sql = "SELECT * FROM messages WHERE toUser=\"$username\" AND room = \"$chatroom\"";
		$result = mysql_query($sql);
		$onResult = 0;
		$usernameDisplay = ""; //How the username is displayed
		$allMessages = array();
		
		#<message_separator> seperates the different messages, while <time_separator> separates the timestamp and username from the message
		while($row = mysql_fetch_array($result, MYSQL_ASSOC)){
			if ($row["type"] == 'first_person'){ //To allow for /me to show as a "first persosn" message
				//Display a "first person" message
				$usernameDisplay = $row["fromUser"]." ";
			}
			else{
				//Display the messasge regularly, with a timestamp
				$usernameDisplay = "#".$row["fromUser"]." (".date("g:i.s a", strtotime($row["time"]))."): ";
			}
			
			//Add message to the array that will be sent back to the user
			array_push($allMessages, $usernameDisplay."<time_separator>".$row["id"]."<time_separator>".$row["message"]);

			$onResult++;
		}
		
		mysql_free_result($result);
		
		//Recipts are in the form  of a comma separated string of message IDs
		//This deletes all the messages with the ID corresponding to a recipt
		if($recipts != "None"){
			foreach(explode(",", $recipts) as $recipt){
				$sql = "DELETE FROM messages WHERE id = $recipt";
				$result = mysql_query($sql);
			}
		}
		//Send the requested messages to the user, separated by a distinct identifier. The identifiers long length will prevent
		//The encryption algorithm from accidently messing up the results
		return implode("<message_separator>", $allMessages);
	}
}


?>

Python_Client/IMCrypt.py

#  IMSafe encryption software
#  Copyright (C) 2014 Patrick T. Cossette
#  www.DigitalDiscrepancy.com
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import random
import hashlib

def encryptText(text, key):
	rand = chr(random.randint(1, 255))
	result = ""
	onChar = 0
	
	hash = hashlib.sha512()
	hash.update(rand+key)
	hash = hash.hexdigest()
	ary = []
	hashLength = len(hash)

	for i in range(0, hashLength, 2):
		ary.append(int(hash[i:i+2], 16))
	
	hashLength = len(ary)
	
	for char in text:
		if onChar > hashLength-1:onChar = 0
		num = ord(char)
		num += ary[onChar]
		if num > 256:
			num -= 256
		
		result += chr(256-num)
		onChar += 1

	return rand + result
	
def decryptText(text, key):
	result = ""
	onChar = 0
	rand = text[0]
	text = text[1::]
	
	hash = hashlib.sha512()
	hash.update(rand+key)
	hash = hash.hexdigest()
	ary = []
	hashLength = len(hash)
	
	for i in range(0, hashLength, 2):
		ary.append(int(hash[i:i+2], 16))
	
	hashLength = len(ary)
	
	
	for char in text:
		if onChar > hashLength-1:onChar = 0
		num = ord(char)
		num = 256-num
		num -= ary[onChar]
		if num < 0:
			num += 256
		
		result += chr(num)
		onChar += 1

	return result
	


Python_Client/IMCrypt2.py


#  IMSafe encryption software
#  Copyright (C) 2014 Patrick T. Cossette
#  www.DigitalDiscrepancy.com
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import random
import hashlib
import base64

def encryptText(text, key):
        rand = chr(random.randint(1, 255))
        result = ""
        onChar = 0
        
        hash = hashlib.sha512()
        hash.update(rand+key)
        hash = hash.hexdigest()
        ary = []
        hashLength = len(hash)
        
        for i in range(0, hashLength, 2):
            ary.append(int(hash[i:i+2], 16))
    
        hashLength = len(ary)
        
        for char in text:
            if onChar > hashLength-1:onChar = 0
            num = ord(char)
            num += ary[onChar]
            if num > 256:
                num -= 256
        
            result += chr(256-num)
            onChar += 1

        return base64.b64encode(rand + result)

def decryptText(text, key):
    
        text = base64.b64decode(text)
        
        result = ""
        onChar = 0
        rand = text[0]
        text = text[1::]
        
        hash = hashlib.sha512()
        hash.update(rand+key)
        hash = hash.hexdigest()
        ary = []
        hashLength = len(hash)
        
        for i in range(0, hashLength, 2):
            ary.append(int(hash[i:i+2], 16))
        
        hashLength = len(ary)
        
        
        for char in text:
            if onChar > hashLength-1:onChar = 0
            num = ord(char)
            num = 256-num
            num -= ary[onChar]
            if num < 0:
                num += 256
            
            result += chr(num)
            onChar += 1

        return result



Python_Client/IMNetwork.py


#  IMSafe encryption software
#  Copyright (C) 2014 Patrick T. Cossette
#  www.DigitalDiscrepancy.com
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import socket, urllib, urllib2, threading, time


class IMNetwork:
	def __init__(self):
		self.ip = None
		self.port = 80
		self.socket = None
		
		self.connected = False
		self.screen_name = ""
		self.address = 'http://www.example.com/IMSafe.php'
		self.updateCallback = self.dummy
		self.recipts = []
		self.count =1
		
	def dummy(self, msg):
		pass
		
	def list(self):
		result = self.sendData("list", prg = True, msg = "list")
		if not result == "":
			self.updateCallback.__call__("list"+result)
			
	def update(self):
		if len(self.recipts):
			result = self.sendData("", prg = True, msg = "update", delRecipts = ",".join(self.recipts))
			self.recipts = []
			
		else:
			result = self.sendData("", prg = True, msg = "update", delRecipts = "None")
			
		self.recipts = []
		if not result == "" and self.connected:
			self.updateCallback.__call__("update"+result)
			
			
	def updateLoop(self, thread):
		while True:
			if not self.connected:
				break
				
			time.sleep(2);
			self.update()
				
		
	def sendData(self, data, prg = False, msg = None, delRecipts = None, options = None):
		page = ""
		path = ""

		try:
			data = [('username', self.screen_name), ('message', data)]
			if prg:
				data.append(("prg", msg))
			if delRecipts:
				data.append(("recipts", delRecipts))
			if options:
				data.append(("options", options))
				
			data = urllib.urlencode(data)

			if self.address[0:7] != "http://":
				self.address = "http://" + self.address
				
			path = self.address
			req  = urllib2.Request(path, data)
			req.add_header("Content-type", "application/x-www-form-urlencoded")
			page = urllib2.urlopen(req).read()

			
		except urllib2.URLError, error:
			return (0, error)
			self.connected = False
			
		print "page: ",page
		
		if not self.connected:
			if page == "Connection Successful":
				self.connected = True
				threading._start_new_thread(self.updateLoop, ("Update Thread",))
				return  (1, "Connection Successful")
				
			else:
				return (0, "Connection Error!\nPossible Causes:\n1. You are not connected to the internet\n2. You mistyped the server URL\n3. You aren't on the server's whitelist\n4. Your username is blacklisted")
				
		else:
			if page == "BANHAMMER":
				self.connected = False
				return (0, "banned")
				
			elif page == "KICKED":
				self.connected = False
				return (0, page)
				
			elif prg:
				return page
			
			else:
				return (2, page)
		
	def disconnect(self):
		self.connected = False
		result = self.sendData("", prg = True, msg = "disconnect")
		
	def connect(self):
		result = self.sendData("connect_request")
		return result

		
		
